### PR TITLE
Add distributed restore point functionality

### PR DESCRIPTION
**Pull request by pmwkaa**
This change adds `create_distributed_restore_point()` function
which allows to create recovery restore point across
nodes.

Issue: https://github.com/timescale/timescaledb/issues/2846

### CODE REVIEW

File: tsl/src/backup.c [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+	if (RecoveryInProgress())
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				 errmsg("restore point cannot be created during recovery")));
+
+	/* create restore point locally and on each data node */
+	lsn = XLogRestorePoint(name);
+
+	sql = psprintf("SELECT pg_catalog.pg_create_restore_point(%s)", quote_literal_cstr(name));
```
Comment (erimatnor) [inline tsl/src/backup.c:59]: Maybe we should do:

```suggestion
	sql = psprintf("SELECT pg_create_restore_point AS lsn FROM pg_catalog.pg_create_restore_point(%s)", quote_literal_cstr(name));
```

**Review by erimatnor — COMMENTED** [review]
Some quick feedback.

File: sql/ddl_api.sql [hunk 2]
```diff
@@ -194,6 +194,12 @@ CREATE OR REPLACE PROCEDURE distributed_exec(
        transactional BOOLEAN = TRUE)
 AS '@MODULE_PATHNAME@', 'ts_distributed_exec' LANGUAGE C;
 
+-- Execute create_restore_point() on each data node
+CREATE OR REPLACE FUNCTION create_distributed_restore_point(
+    name                     NAME
```
[the remark this answers is not in the document]
Comment (pmwkaa) [reply sql/ddl_api.sql:199]: Yes, but restore point name has also limit by name size (which is 64 chars max). I can switch to TEXT and add additional check for length in such case

File: sql/ddl_api.sql [hunk 3]
```diff
@@ -194,6 +194,12 @@ CREATE OR REPLACE PROCEDURE distributed_exec(
...
+) RETURNS pg_lsn
```
[the remark this answers is not in the document]
Comment (pmwkaa) [reply sql/ddl_api.sql:200]: This was my original proposal, but I think we decided not to do when we discussed it. I am happy to try it out then, makes sense to me

[the remark this answers is not in the document]
Comment (pmwkaa) [reply sql/ddl_api.sql:200]: We will also have to return AN's lsn, which can't be described as a node. I guess we could set node name to NULL in such case

File: sql/ddl_api.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply sql/ddl_api.sql]: I think we should use same type as `pg_create_restore_point`, which uses `TEXT`. As put in the PostgreSQL docs: `The name type exists only for the storage of identifiers in the internal system catalogs and is not intended for use by the general user.`. Since the restore point is not a PG identifier I think `TEXT` is appropriate here.

File: sql/ddl_api.sql [hunk 4]
```diff
@@ -194,6 +194,12 @@ CREATE OR REPLACE PROCEDURE distributed_exec(
        transactional BOOLEAN = TRUE)
 AS '@MODULE_PATHNAME@', 'ts_distributed_exec' LANGUAGE C;
 
+-- Execute pg_create_restore_point() on each data node
+CREATE OR REPLACE FUNCTION create_distributed_restore_point(
+    name                   TEXT
+) RETURNS TABLE(node_name NAME, restore_point pg_lsn)
```
Comment (feikesteenbergen) [inline sql/ddl_api.sql:200]: Currently, in the output `node_name IS NULL` implies that it is the access node.
What if we'd be more explicit about this and return an additional column 

`(node_name NAME, node_role TEXT NOT NULL, restore_point pg_lsn)`

OTOH, this is exactly how some pg catalogs also work (`pg_db_role_setting` which if `setrole IS NULL` means it is set for the whole database), so there's some precedent here.

**Comment by feikesteenbergen** [comment]
What is the mechanism that will ensure we have a consistent backup across the whole topology?

I don't see anything here that ensures all data nodes and the access node is on the same page - even though it would be intrusive - it might be best to create a synchronization point between all members of the cluster before creating the restore point, and somehow return the timestamp at which consistency was ensured across the board.

**Comment by feikesteenbergen** [comment]
This looks good, we can use `pgBackRest` fine with these restore points, so this should give us enough functionality to be able to offer recovery for multinode to distinct points in time.

File: tsl/src/backup.c [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+		if (dist_util_membership() != DIST_MEMBER_ACCESS_NODE)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+					 errmsg("distributed restore point must be created on the access node"),
+					 errhint("pg_create_restore_point() can be used to to create restore point "
+							 "on the data node.")));
+
+		/* create local restore point and on each data node */
+		lsn = XLogRestorePoint(name);
```
Comment (erimatnor) [inline tsl/src/backup.c:89]: So, I believe you have to explicitly lock the `remote_txn` table before creating the restore point, because we need to ensure that there aren't any concurrent distributed transactions that have committed and will become visible to us after the restore point.

Comment (erimatnor) [reply tsl/src/backup.c:89]: And, we need to ensure that other transactions hold their share lock until the end of their transactions.

Comment (erimatnor) [reply tsl/src/backup.c:114]: Also, perhaps it is better to take the local (AN) restore point after all the data node have taken restore points? While locking can ensure that it is kind of "atomic" I think it makes sense to have AN last. Thoughts?

File: tsl/test/expected/backup.out [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+DROP DATABASE dist_rp_test;
+-- test on data node
+\c :MY_DB1 :ROLE_CLUSTER_SUPERUSER;
+\set ON_ERROR_STOP 0
+SELECT create_distributed_restore_point('test');
+ERROR:  distributed restore point must be created on the access node
+\set ON_ERROR_STOP 1
+\c :TEST_DBNAME :ROLE_CLUSTER_SUPERUSER;
+-- test on access node
+SELECT node_name, pg_lsn(restore_point) > pg_lsn('0/0') FROM create_distributed_restore_point('dist_rp');
```
Comment (erimatnor) [inline tsl/test/expected/backup.out:89]: Should probably add a column alias, although I am not sure what a suitable name for the condition is (I realize it is for making tests reproducible):

```suggestion
SELECT node_name, pg_lsn(restore_point) > pg_lsn('0/0') AS valid_lsn FROM create_distributed_restore_point('dist_rp');
```

File: tsl/test/expected/backup.out [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  node_name  | ?column? 
+-------------+----------
+             | t
+ data_node_1 | t
+ data_node_2 | t
+ data_node_3 | t
+(4 rows)
+
+-- restore point can be have duplicates
+SELECT node_name, pg_lsn(restore_point) > pg_lsn('0/0') FROM create_distributed_restore_point('dist_rp');
```
Comment (erimatnor) [inline tsl/test/expected/backup.out:99]: Same here as above.

File: tsl/test/expected/backup.out [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+SELECT restore_point as lsn_2 FROM create_distributed_restore_point('dist_rp_2') WHERE node_name IS NULL \gset
+SELECT pg_lsn(:'lsn_2') > pg_lsn(:'lsn_1');
+ ?column? 
+----------
+ t
+(1 row)
+
+-- make sure it is compatible with local restore point
+SELECT pg_create_restore_point('dist_rp') as lsn_3 \gset
+SELECT pg_lsn(:'lsn_3') > pg_lsn(:'lsn_2');
```
Comment (erimatnor) [inline tsl/test/expected/backup.out:119]: Same here

File: tsl/test/expected/backup.out [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+ f
+(1 row)
+
+-- test on single node
+CREATE DATABASE dist_rp_test;
+\c dist_rp_test :ROLE_CLUSTER_SUPERUSER
+SET client_min_messages TO ERROR;
+CREATE EXTENSION timescaledb;
+\set ON_ERROR_STOP 0
+SELECT create_distributed_restore_point('test');
```
Comment (erimatnor) [inline tsl/test/expected/backup.out:76]: What happens if this is called via `distributed_exec` from the access node?

File: tsl/test/expected/backup.out [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+ERROR:  distributed restore point must be created on the access node
+\set ON_ERROR_STOP 1
+\c :TEST_DBNAME :ROLE_CLUSTER_SUPERUSER;
+DROP DATABASE dist_rp_test;
```
Comment (erimatnor) [inline tsl/test/expected/backup.out:80]: Maybe use `distributed_exec` to avoid creating new session?

File: tsl/test/expected/backup.out (hunk 9 above)
Comment (pmwkaa) [reply tsl/test/expected/backup.out:77]: It will fail as expected, added the case

File: sql/ddl_api.sql (hunk 4 above)
Comment (pmwkaa) [reply sql/ddl_api.sql:200]: I think this is fine, I'll bring this point to discussion with everyone

**Comment by pmwkaa** [comment]
> What is the mechanism that will ensure we have a consistent backup across the whole topology?
> 
> I don't see anything here that ensures all data nodes and the access node is on the same page - even though it would be intrusive - it might be best to create a synchronization point between all members of the cluster before creating the restore point, and somehow return the timestamp at which consistency was ensured across the board.

Yes, synchronization logic is there now according to the design doc with new isolation tests

**Review by feikesteenbergen — APPROVED** [review]

File: tsl/src/backup.c [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+		 * We must avoid possible situation when a concurrent remote transaction
+		 * commits on the access node during the creation of the distributed
+		 * restore point and gets into the data nodes WAL before the
+		 * restore point marker written.
+		 *
+		 * To do that we take an exclusive lock on the remote transaction
+		 * table, which will force any concurrent transaction
+		 * wait during their PREPARE phase.
+		 */
+		LockRelationOid(ts_catalog_get()->tables[REMOTE_TXN].id, ExclusiveLock);
```
Comment (erimatnor) [inline tsl/src/backup.c:113]: I guess this lock level is OK, but, FWIW, the minimum lock level to conflict with other distributed transactions is `RowExclusiveLock`. Do we really need to take anything heavier than that? (i.e., this blocks things like VACUUM, etc.)

**Review by erimatnor — COMMENTED** [review]
I am a little bit unsure of the locking here.

In particular, we want the lock we take in `create_distributed_restore_point` to conflict with the `RowExclusiveLock` taken when writing to the remote TX log, which happens in  `remote_txn_persistent_record_write`. However, in that function we release the lock as soon as we've written the entry. We should probably make sure that it is held to the end of the transaction there so that we know in `create_distributed_restore_point` that, once we have the lock, there are no other distributed transactions that can send `COMMIT PREPARED` from that point on.

**Comment by pmwkaa** [comment]
> I am a little bit unsure of the locking here.
> 
> In particular, we want the lock we take in `create_distributed_restore_point` to conflict with the `RowExclusiveLock` taken when writing to the remote TX log, which happens in `remote_txn_persistent_record_write`. However, in that function we release the lock as soon as we've written the entry. We should probably make sure that it is held to the end of the transaction there so that we know in `create_distributed_restore_point` that, once we have the lock, there are no other distributed transactions that can send `COMMIT PREPARED` from that point on.

Not sure I follow. Since we release lock in `create_distributed_restore_point()` when transaction completes we control the critical section there and it is exclusive. It will block other lock waiters and it actually not matter when they release lock, because they will be put on wait during the lock taking in the first place.

File: sql/ddl_api.sql [hunk 12]
```diff
@@ -194,6 +194,12 @@ CREATE OR REPLACE PROCEDURE distributed_exec(
        transactional BOOLEAN = TRUE)
 AS '@MODULE_PATHNAME@', 'ts_distributed_exec' LANGUAGE C;
 
+-- Execute pg_create_restore_point() on each data node
+CREATE OR REPLACE FUNCTION create_distributed_restore_point(
```
Comment (k-rus) [inline sql/ddl_api.sql:198]: Is there a test, which calls this function and shows its result? I see it is called in the isolation test, but I don't see its output? Can you point or add such test?

Comment (pmwkaa) [reply sql/ddl_api.sql:198]: Result of this function is not deterministic or prone to be flaky, it is called like this  https://github.com/timescale/timescaledb/pull/2843/files#diff-760d95c35de4c932eb2777985ff2b367dc10a03a548931171becbe04ab81d76cR94 to be more stable in tests

File: tsl/src/backup.c (hunk 11 above)
Comment (pmwkaa) [reply tsl/src/backup.c:113]: I guess it is fine to downgrade it as long as it does what it is intended to, I'll give it a try

**Review by k-rus — COMMENTED** [review]
Just few minor comments. I haven't looked further, so I can review after Erik's comments are addressed.

File: tsl/src/backup.c [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+static Datum
+create_restore_point_datum(TupleDesc tupdesc, const char *node_name, XLogRecPtr lsn)
+{
+	Datum values[_Anum_restore_point_max];
+	bool nulls[_Anum_restore_point_max] = { false };
+	HeapTuple tuple;
+
+	tupdesc = BlessTupleDesc(tupdesc);
+	values[AttrNumberGetAttrOffset(Anum_restore_point_node_name)] = CStringGetDatum(node_name);
```
[the remark this answers is not in the document]
Comment (pmwkaa) [reply tsl/src/backup.c:40]: The idea of this line was to set NULL in value as well without adding it under if, since CStringGetDatum does not do any transformations. I guess I can make it more clear as you suggest

**Comment by erimatnor** [comment]
> > I am a little bit unsure of the locking here.
> > In particular, we want the lock we take in `create_distributed_restore_point` to conflict with the `RowExclusiveLock` taken when writing to the remote TX log, which happens in `remote_txn_persistent_record_write`. However, in that function we release the lock as soon as we've written the entry. We should probably make sure that it is held to the end of the transaction there so that we know in `create_distributed_restore_point` that, once we have the lock, there are no other distributed transactions that can send `COMMIT PREPARED` from that point on.
> 
> Not sure I follow. Since we release lock in `create_distributed_restore_point()` when transaction completes we control the critical section there and it is exclusive. It will block other lock waiters and it actually not matter when they release lock, because they will be put on wait during the lock taking in the first place.

I matters because if the lock is released before the transaction ends, the result of the transaction won't yet be visible to the next transaction that gets the lock. If the the lock in  `remote_txn_persistent_record_write` is held until transaction end, we know that, when the lock is grabbed in `create_distributed_restore_point()`, all other concurrently running 2PC transactions have committed their changes (to, e.g., the `remote_txn` table) and made them visible.

File: tsl/src/backup.c (hunk 11 above)
Comment (erimatnor) [reply tsl/src/backup.c:119]: Correction: the minimum lock level to conflict I believe would be `ShareRowLock`. So, although `ExclusiveLock` certainly works, it is strictly higher than necessary. Perhaps this doesn't matter for this case, though, since currently we know the processes involved. Anyway, do what you deem best here.

File: tsl/src/backup.c [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+		int name_len = strlen(name);
+		MemoryContext oldctx;
+		TupleDesc tupdesc;
+		char *sql;
+
+		if (name_len >= MAXFNAMELEN)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("restore point name is too long"),
+					 errhint("maximum length is %d, while provided name has %d chars",
```
Comment (erimatnor) [inline tsl/src/backup.c:78]: I think this sounds more like "detail" than "hint". A hint should be a concrete suggestion how to avoid the error, e.g., "Use a name with a maximum length of %d.". You can of course reformulate like that if you prefer a hint (often it is actually more helpful to tell users what to do than pointing out the details of why their command failed). It is a bit pointless, IMO, to point out the length of the restore point name since it was suggested by the user, so the hint message above might be more suitable.

If you prefer to keep the detail message, then tweak to conform to error guidelines:

```suggestion
					 errdetail("Maximum length is %d, while provided name has %d chars.",
```

File: tsl/src/backup.c [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+
+		/* create local restore point and on each data node */
+		lsn = XLogRestorePoint(name);
+
+		sql = psprintf("SELECT pg_create_restore_point AS lsn "
+					   "FROM "
+					   "pg_catalog.pg_create_restore_point(%s)",
+					   quote_literal_cstr(name));
+
+		result_cmd = ts_dist_cmd_invoke_on_all_data_nodes(sql);
```
Comment (erimatnor) [inline tsl/src/backup.c:143]: Is this correct? It looks like you are saving the result_cmd in the SRF multi-call context, but not in the corresponding memory context.

File: tsl/src/backup.c [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+		if (!superuser())
+			ereport(ERROR,
+					(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					 errmsg("must be superuser to create restore point")));
+
+		if (!ts_guc_enable_2pc)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+					 errmsg("two-phase commit transactions are not enabled"),
+					 errhint("timescale.enable_2pc must be set to true.")));
```
Comment (erimatnor) [inline tsl/src/backup.c:105]: Note wrong prefix:

```suggestion
					 errhint("Set timescaledb.enable_2pc to true.")));
```

File: tsl/src/backup.c [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+
+		if (dist_util_membership() != DIST_MEMBER_ACCESS_NODE)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+					 errmsg("distributed restore point must be created on the access node"),
+					 errhint("pg_create_restore_point() can be used to create restore point "
+							 "on the data node.")));
```
Comment (erimatnor) [inline tsl/src/backup.c:112]: I am not sure we need this hint because it can be misleading and it suggests a solution that doesn't achieve the original goal, i.e., to create a distributed restore point. If the user tries to do X we shouldn't assume the user was really trying to do Y and direct them towards doing that instead. Further, a standard restore point on a data node isn't very useful so we should assume that the user simply aimed to create a distributed restore point but executed it on the wrong node. Either remove the hint or change it to something like:

```suggestion
					 errhint("Connect to the access node and create the distributed restore point from there.")));
```

File: tsl/test/isolation/expected/dist_restore_point.out [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+               
+step s1_create_dist_rp: SELECT restore_point > pg_lsn('0/0') as valid_lsn FROM create_distributed_restore_point('s1_test'); <waiting ...>
+step s2_insert: INSERT INTO disttable VALUES ('2019-08-02 10:45', 0, 0.0); <waiting ...>
+step s3_lock_count: 
+	SELECT waitpoint_locks('create_distributed_restore_point_lock') as cdrp_locks, 
+	       remote_txn_locks() as remote_txn_locks;
+
+cdrp_locks     remote_txn_locks
+
+2              2
```
Comment (erimatnor) [inline tsl/test/isolation/expected/dist_restore_point.out:40]: What's expected in terms of lock counts here (maybe add in comment). I'd expect 2 cdrp locks (one for the lock session and one for the create-restore-point session). But why do we see 2 remote_txn locks when there is only one INSERT session?

Would be useful to break this down in waiters and holders (although I guess there is always 1 holder).

File: tsl/test/isolation/expected/dist_restore_point.out [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+data_node_3    localhost      55432          cdrp_3         f              f              f              
+step s2_begin: BEGIN;
+step s2_update: UPDATE disttable SET temp = temp + 1.0;
+step s3_lock_enable: SELECT debug_waitpoint_enable('create_distributed_restore_point_lock');
+debug_waitpoint_enable
+
+               
+step s1_create_dist_rp: SELECT restore_point > pg_lsn('0/0') as valid_lsn FROM create_distributed_restore_point('s1_test'); <waiting ...>
+step s2_commit: COMMIT; <waiting ...>
+step s3_lock_count:
```
Comment (erimatnor) [inline tsl/test/isolation/expected/dist_restore_point.out:72]: Don't we want to show the lock counts before COMMIT? I.e., we want to see that the restore point session blocks on the remote_txn lock instead of the waitpoint lock.

File: tsl/test/isolation/specs/dist_restore_point.spec.in [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+}
+step "s3_lock_enable"   { SELECT debug_waitpoint_enable('create_distributed_restore_point_lock'); }
+step "s3_lock_release"  { SELECT debug_waitpoint_release('create_distributed_restore_point_lock'); }
+step "s3_lock_count"    {
+	SELECT waitpoint_locks('create_distributed_restore_point_lock') as cdrp_locks, 
+	       remote_txn_locks() as remote_txn_locks;
+}
+step "s3_lock_count_fs" { SELECT foreign_server_locks() as foreign_server_locks; }
+
+permutation "s2_setup"
```
[the remark this answers is not in the document]
Comment (pmwkaa) [reply tsl/test/isolation/specs/dist_restore_point.spec.in:85]: I can't recall exactly, but I believe it has something to do with the limitations, it was likely problematic to set this up during session creation, maybe some of the functions was not transactional

File: tsl/test/isolation/expected/dist_restore_point.out (hunk 18 above)
Comment (pmwkaa) [reply tsl/test/isolation/expected/dist_restore_point.out:40]: Because first lock is already taken by the create_distributed_restore_point() function. Basically waitpoint is reached after this lock being taken

File: tsl/test/isolation/expected/dist_restore_point.out (hunk 19 above)
Comment (pmwkaa) [reply tsl/test/isolation/expected/dist_restore_point.out:72]: This is what it does, because waitpoint lock is taken after the remote_txn lock, we can see them both. I've updated the test to see those counters before the commit

File: tsl/src/remote/txn.c [hunk 21]
```diff
@@ -754,7 +754,9 @@ remote_txn_persistent_record_write(TSConnectionId cid)
 
 	rel = table_open(catalog->tables[REMOTE_TXN].id, RowExclusiveLock);
 	persistent_record_insert_relation(rel, id);
-	table_close(rel, RowExclusiveLock);
 
+	/* Keep the table lock until transaction completes in order to
+	 * synchronize with distributed restore point creation */
+	table_close(rel, NoLock);
```
Comment (k-rus) [inline tsl/src/remote/txn.c:760]: Does it mean that all distributed transactions will be serialized due to keeping this lock until the end? cc @pmwkaa @erimatnor

File: tsl/test/sql/CMakeLists.txt [hunk 22]
```diff
@@ -60,6 +60,7 @@ set(TEST_FILES_DEBUG
   transparent_decompression_queries.sql
   tsl_tables.sql
   license.sql
+  backup.sql
```
Comment (k-rus) [inline tsl/test/sql/CMakeLists.txt:63]: Since the test is about distributed case, I suggest to follow other distributed tests and prefix with `dist_`.
```suggestion
  dist_backup.sql
```
Can it be also moved into the place with other distributed tests, i.e., follow lexicographical order.

File: tsl/test/sql/backup.sql [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+SET ROLE :ROLE_1;
+SELECT create_distributed_restore_point('test');
+RESET ROLE;
+\set ON_ERROR_STOP 1
+
+-- make sure 2pc are enabled
+SET timescaledb.enable_2pc = false;
+\set ON_ERROR_STOP 0
+SELECT create_distributed_restore_point('test');
+\set ON_ERROR_STOP 1
```
Comment (k-rus) [inline tsl/test/sql/backup.sql:60]: I don't follow how 2PC is enabled back. Is it reset automatically after one transaction? Can you do `SHOW timescaledb.enable_2pc`, so it is clear that it is enabled back?

**Review by k-rus — COMMENTED** [review]
My main worry is that keeping the lock until the end of each distributed transaction will serialize all distributed transaction and will not allow to execute them in parallel. Am I correct or why is it not an issue?

How is it difficult to check that the restore points were created in the WAL? 

I have also number of non-critical comments and suggestions for improvements.

File: tsl/src/remote/txn.c (hunk 21 above)
Comment (pmwkaa) [reply tsl/src/remote/txn.c:760]: To my understanding this lock should not affect other distributed transaction accessing this table, because this lock would block only on conflicting key updates, which should not ever happen. We are locking this table with higher priority lock in create distributed restore point function, so basically it will lock other sessions untill that function completes

Comment (k-rus) [reply tsl/src/remote/txn.c:760]: Isn't it, every transaction calls this function insert a record into it? The lock locks entire the remote transaction table, isn't it?
Since the function is called before sending the prepare to data nodes, the lock will be kept while negotiating prepare until the commit is sent.

File: tsl/src/backup.c [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+					 (errmsg("recovery is in progress"),
+					  errhint("WAL control functions cannot be executed during recovery."))));
+
+		if (!XLogIsNeeded())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 errmsg("WAL level '%s' is not sufficient for creating a restore point",
+							GetConfigOptionByName("wal_level", NULL, false)),
+					 errhint(
+						 "wal_level must be set to \"replica\" or \"logical\" at server start.")));
```
[the remark this answers is not in the document]
Comment (pmwkaa) [reply tsl/src/backup.c:94]: Just to note, that this check the check above with message are taken from PostgreSQL implementation of pg_create_restore_point function

File: tsl/test/sql/backup.sql (hunk 23 above)
Comment (pmwkaa) [reply tsl/test/sql/backup.sql:60]: Good catch, likely accidentally removed it while refactoring

Comment (k-rus) [reply tsl/test/sql/backup.sql:60]: Then why have the regression test succeeded?

File: tsl/src/remote/txn.c (hunk 21 above)
Comment (pmwkaa) [reply tsl/src/remote/txn.c:760]: Not its not, to my understanding this lock does not lock other transactions, unless they have conflicting keys. Otherwise we would also use it in the create_distributed_restore_point function, attempt to use it was unsuccessful

Comment (k-rus) [reply tsl/src/remote/txn.c:760]: To my understanding `rel = table_open(catalog->tables[REMOTE_TXN].id, RowExclusiveLock);` locks entire table. No keys to specific transactions are used here.

File: tsl/test/sql/backup.sql (hunk 23 above)
Comment (pmwkaa) [reply tsl/test/sql/backup.sql]: Because we reconnected in tests after that and it set it to the default value

Comment (k-rus) [reply tsl/test/sql/backup.sql]: Thank you for clarifying! Basically it became different session.

File: tsl/src/remote/txn.c (hunk 21 above)
Comment (pmwkaa) [reply tsl/src/remote/txn.c:760]: Again, it's not. Row-level locks do not affect data querying, they block only writers to the same row.

Comment (k-rus) [reply tsl/src/remote/txn.c:760]: Actually it is a [table level lock](https://www.postgresql.org/docs/12/explicit-locking.html#LOCKING-TABLES). However it doesn't conflict with itself, so I guess it conflicts only on the same row. It wasn't obvious to me, so you are right that it should not be an issue. Thank you for the patience.

**Comment by k-rus** [comment]
> How is it difficult to check that the restore points were created in the WAL?

@pmwkaa Or will it be handled separately?

**Review by k-rus — APPROVED** [review]

**Comment by pmwkaa** [comment]
> > How is it difficult to check that the restore points were created in the WAL?
> 
> @pmwkaa Or will it be handled separately?

I couldn't find an easy way to do this using our test suite, also checked native PostgreSQL tests. I guess we could cover it in other issue related to tooling, etc

File: tsl/src/backup.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply tsl/src/backup.c]: If you want to keep for consistency with PG, then that's fine. Otherwise I think the suggested change is a nicer message. :)

File: tsl/src/backup.c [hunk 25]
```diff
@@ -0,0 +1,192 @@
...
+					 errmsg("restore point name is too long"),
+					 errdetail("Maximum length is %d, while provided name has %d chars.",
+							   MAXFNAMELEN - 1,
+							   name_len)));
+
+		if (RecoveryInProgress())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 (errmsg("recovery is in progress"),
+					  errhint("WAL control functions cannot be executed during recovery."))));
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply tsl/src/backup.c]: I agree it sounds like a detail message.

File: tsl/src/dist_backup.c [hunk 26]
```diff
@@ -0,0 +1,194 @@
...
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 (errmsg("recovery is in progress"),
+					  errdetail("WAL control functions cannot be executed during recovery."))));
+
+		if (!XLogIsNeeded())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 errmsg("WAL level '%s' is not sufficient for creating a restore point",
+							GetConfigOptionByName("wal_level", NULL, false)),
+					 errhint("Set wal_level must to \"replica\" or \"logical\" at server start.")));
```
Comment (erimatnor) [inline tsl/src/dist_backup.c:96]:
```suggestion
					 errhint("Set wal_level to \"replica\" or \"logical\" at server start.")));
```

File: tsl/src/dist_backup.c [hunk 27]
```diff
@@ -0,0 +1,194 @@
...
+
+		/*
+		 * In order to achieve synchronization across the multinode cluster,
+		 * we must ensure that the restore point created on the access node is
+		 * synchronized with each data node.
+		 *
+		 * We must avoid possible situation when a concurrent remote transaction
+		 * commits on the access node during the creation of the distributed
+		 * restore point and gets into the data nodes WAL before the
+		 * restore point marker written.
```
Comment (erimatnor) [inline tsl/src/dist_backup.c:124]: I found the comment a bit hard to follow. Suggest something like:
```suggestion
		 * We must ensure that no concurrent prepared transactions are 
		 * committed (COMMIT PREPARED) while we create the restore point.
		 * Otherwise, the distributed restore point might include prepared transactions
		 * that have committed on some data nodes but not others, leading to an
		 * inconsistent state when the distributed database is restored from a backup
		 * using the restore point.
```

File: tsl/test/isolation/specs/dist_restore_point.spec.in (hunk 20 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply tsl/test/isolation/specs/dist_restore_point.spec.in:85]: We probably shouldn't have something there that we don't know the reason for. It seems to me  that this does nothing so can we remote it?

**Review by erimatnor — APPROVED** [review]
Approving, but added a few nits/suggestions. In particular, I don't think we should have that initial permutation in the test. It seems to have no purpose and if we cannot explain why it is there, we shouldn't have it there.

File: sql/ddl_api.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (feikesteenbergen) [reply sql/ddl_api.sql]: Would also be good to return the current time, as we need the time of the restore point to find the previous backup.
